java-backend/src/main/java/org/kframework/backend/java/util/RuleSourceUtil.java

```diff
@@ -0,0 +1,59 @@
+// Copyright (c) 2015-2018 K Team. All Rights Reserved.
+package org.kframework.backend.java.util;
+
+import org.kframework.attributes.Location;
+import org.kframework.backend.java.kil.Rule;
+import org.kframework.utils.file.FileUtil;
+
+import java.io.File;
+import java.util.Collections;
+import java.util.Formatter;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * @author Denis Bogdanas
+ * Created on 27-Aug-18.
+ */
+public class RuleSourceUtil {
+    private static final Map<Rule, String> cache = Collections.synchronizedMap(new HashMap<>());
+
+    public static boolean sourceShortEnough(Rule rule) {
+        Location location = rule.getLocation();
+        return location != null && location.endLine() - location.startLine() <= 20;
+    }
+
+    public static String loadSource(Rule rule) {
+        if (!cache.containsKey(rule)) {
+            if (rule.getSource() != null && rule.getLocation() != null) {
+                cache.putIfAbsent(rule, FileUtil.loadFragment(new File(rule.getSource().source()), rule.getLocation()));
+            }
+        }
+        return cache.get(rule);
+    }
+
+    public static void printSourceAndRule(Rule rule) {
+        File source = rule.source().isPresent() ? new File(rule.getSource().source()) : null;
+        System.err.format("\nSource: %s %s\n", source, rule.getLocation());
+        if (sourceShortEnough(rule)) {
+            System.err.println(loadSource(rule));
+        } else {
+            System.err.println("rule too long or location unknown...");
+        }
+    }
+
+    public static void printRuleAndSource(Rule rule) {
```

Comment:
having both a printSourceAndRule and printRuleAndSource is really confusing, especially when one calls another helper function that has a lot of duplicated code. Did you really intend to keep both around?

Reply:
Yes, to have nice to read error messages I need both of them in K-gnosis. And I don;t want to have this file different in the 2 branches. Hopefully I'll break most of the stuff in K-gnosis to pieces some day and PR to master.

Reply:
@dwightguth replied and fixed what's possible.

Since this class is only for error reporting, is it really worth testing? It doesn't affect rewrite engine functionality. We are late with deadlines in VSC team.

Reply:
no, it's fine, I was more worried about the corner cases but it seems you've handled them now.